Shell: Add capability to retrieve command return codes.

hsearle opened the pull request:
Currently, zephyr discards all results returned from shell commands, forcing the user to rely on print statements. These changes instead store the return code in the shell context and add a command to retrieve the code for the most recently executed command.

nordic-krch commented on subsys/shell/Kconfig:
```diff
@@ -251,6 +251,13 @@ config SHELL_AUTOSTART
 	  If enabled, shell will be automatically started.
 
+config SHELL_RETURN_VALUE
+	bool "Command to retrieve return value"
```
it should depend on `SHELL_CMDS`. Could also be renamed to `SHELL_CMDS_RETURN_VALUE` as there are already `SHELL_CMDS_RESIZE` or `SHELL_CMDS_SELECT`.

nordic-krch commented on subsys/shell/shell_retval.c:
```diff
@@ -0,0 +1,10 @@
+	return 0;
+}
+
+SHELL_CMD_REGISTER(return_code, NULL, "Gets return code of previous command.", cmd_get_retval);
```
You don't need a new file. If it depends on `SHELL_CMDS` then you can add it to `shell_cmds.c` and use `SHELL_COND_CMD_REGISTER` (or better `SHELL_COND_CMD_ARG_REGISTER` and indicate that there is only 1 fixed argument).

nordic-krch commented on subsys/shell/shell_retval.c:
```diff
@@ -0,0 +1,10 @@
+static int cmd_get_retval(const struct shell *shell, size_t argc, char **argv){
+	ARG_UNUSED(argc);
+	ARG_UNUSED(argv);
+	shell_print(shell,"%d",shell->ctx->return_code);
```
nit: spaces after commas.

nordic-krch commented on subsys/shell/shell.c:
```diff
@@ -999,7 +999,9 @@ static void state_collect(const struct shell *shell)
 				} else {
 					/* Command execution */
-					(void)execute(shell);
+					int ret_val = execute(shell);
```
nit: empty line after variable declaration.

nordic-krch reviewed (commented):
Thanks, nice feature. Just a few comments.

nordic-krch commented:
You also need to update commit message which should start with subsystem prefix `shell: Add ...`. 
See the guideline for commit message: https://docs.zephyrproject.org/3.1.0/contribute/guidelines.html#commit-guidelines

nordic-krch commented on subsys/shell/shell_cmds.c:
```diff
@@ -10,6 +10,7 @@
 #include "shell_vt100.h"
 
 #define SHELL_MSG_CMD_NOT_SUPPORTED	"Command not supported.\n"
+#define SHELL_HELP_RETVAL       "Print return value of most recent command"
```
Please align string to others.

nordic-krch commented on subsys/shell/shell_cmds.c:
```diff
@@ -402,6 +403,14 @@ static int cmd_resize(const struct shell *shell, size_t argc, char **argv)
+{
+	ARG_UNUSED(argc);
+	ARG_UNUSED(argv);
+	shell_print(shell, "%d", shell->ctx->return_code);
```
`ctx->ret_val`

jakub-uC commented:
I do not understand the point of this change in this form. After all, further the user relies on what is printed only by another command. The codes returned by the commands are not standardized so, in my opinion, it is better after all to rely on what the main command prints.

However, if the return value itself is important then maybe instead of a command `cmd_get_retval` that prints the value it is better to extend a shell API by adding function returning `shell->ctx->return_code`?

hsearle commented:
> I do not understand the point of this change in this form. After all, further the user relies on what is printed only by another command. The codes returned by the commands are not standardized so, in my opinion, it is better after all to rely on what the main command prints.
> 
> However, if the return value itself is important then maybe instead of a command `cmd_get_retval` that prints the value it is better to extend a shell API by adding function returning `shell->ctx->ret_val`?

There are several reasons why this change is preferable to the current method:

1. The main motivation is automation. Having a standard interface, i.e. call
command; call retval; use returned value, is preferable to needing to create
separate parsing for each command. Although the individual codes are not standardized
(though many do have sensible return codes), it is easy to differentiate success/
failure. This is not the case currently, as print statements are inconsistent in
how they signify success/failure, as well as what other information they provide.

2. Having the ability to retrieve return codes eases the burden on command
writers to have to write descriptive print statements. Instead, they would need
only return sensible values and document the meaning of each. This is
something that most command writers are already doing, so it eliminates the
need for extra work when adding new commands.

3. The return values already exist and are passed back to the shell. Adding
a command to retrieve the value is almost trivial. Given that it has some
merit, there doesn't seem to be a good reason not to implement.

Additionally, I do not understand your second comment. The point of this new
command is to make return codes available to the command line user, who don't
have direct access to the shell API, only commands. Short of editing the execute
command to always print the return value (which is undesirable, since users may
not be interested in the value and it would just add unwanted traffic), adding
a simple command such as cmd_get_retval is the simplest solution.

jakub-uC commented:
@hsearle 
Sorry for the late response but I did not have access to a computer.

I am not arguing the point of using return code. It is indisputable and your change is very valuable. Until now, using return codes was basically only for unit tests by calling the `shell_execute_cmd` function. 

Again I stress your change is very valuable just thought it would be more universal to add a new API function to shell that would read the value returned by the last executed command. This would give other commands a clean way to use this information. Your `cmd_get_retval` command, instead of reaching into the shell context directly, should, in my opinion, use an API function to do so.

jakub-uC commented on include/zephyr/shell/shell.h:
```diff
@@ -822,6 +822,9 @@ struct shell_ctx {
 	struct k_mutex wr_mtx;
 	k_tid_t tid;
+
+	/*Store result of previously executed command*/
```
Possibly CI will ask you to add space here and at the end.

nordic-krch commented on include/zephyr/shell/shell.h:
```diff
@@ -1241,6 +1241,15 @@ int shell_obscure_set(const struct shell *shell, bool obscure);
 
+/**
+ * @brief Retrieve return value of most recently executed shell command.
+ * 
```
trailing space after `*`

nordic-krch commented on include/zephyr/shell/shell.h:
```diff
@@ -1241,6 +1241,15 @@ int shell_obscure_set(const struct shell *shell, bool obscure);
+ * @brief Retrieve return value of most recently executed shell command.
+ * 
+ * @param[in] shell Pointer to the shell instance
+ * 
```
trailing space.

This commit is missing signoff and body: `Shell: add retrieve return value via shell API`

carlescufi commented:
> Could someone offer some guidance on the failed checks on compliance and coding guidelines? I understand the trailing whitespace issue. However, the compliance check says that I haven't included a body or a signed off by line in the commit message, both of which I have. Additionally, I got the issue that "tags need to be unique", but my variables are properly scoped, and match other functions. I don't understand these issues, and I don't know how to fix them. Thanks

Regarding the commit formatting, please see here:
https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-message-guidelines

gmarull commented on include/zephyr/shell/shell.h:
```diff
@@ -1238,6 +1241,15 @@ int shell_obscure_set(const struct shell *shell, bool obscure);
+ * 
+ * @retval return value of previous command
+ */
+int shell_get_return_value(const struct shell *shell);
```
this violates MISRA rule 5.7, use `const struct shell *sh` instead.

gmarull replied:
they were likely added before this check was in place

gmarull commented:
please rebase

carlescufi commented:
Argh, there's a conflict now. @hsearle could you please rebase and resolve the conflict?

sjg20 commented on subsys/shell/Kconfig:
```diff
@@ -251,6 +251,14 @@ config SHELL_AUTOSTART
+	depends on SHELL_CMDS
+	default y
+	help
+		Enables a command to retrieve the return value from the previously executed command.
```
Please can you improve this help? What is the command that it enable? What is the range of a return value?

sjg20 replied:
I think 'return code' is a better name than 'return value'. This is special, in that it relates to command execution, not just returning a result from a function.

hsearle replied:
I asked before about the name of the command but received no feedback. If changing it is best, I'd be happy to, but I would appreciate some more input from others whether this is the preferred name.

nordic-krch replied:
Personally, I don't have strong opinion about this. However, what i see is that it would be good to combine shell commands into one root command e.g. `shell` so we have `shell retval` or `shell retcode`. It's not a subject of this PR, though but something that we should consider soon.

sjg20 commented on subsys/shell/shell.c:
```diff
@@ -1001,7 +1001,7 @@ static void state_collect(const struct shell *sh)
 				} else {
 					/* Command execution */
-					(void)execute(sh);
+					shell->ctx->ret_val = execute(shell);
```
s/shell/sh/

Thalley commented:
> Currently, zephyr discards all results returned from shell commands, forcing the user to rely on print statements. These changes instead store the return code in the shell context and add a command to retrieve the code for the most recently executed command.

Has this change taken into account that the shell_cmd_handler is defined as
```c
/**
 * @brief Shell command handler prototype.
 *
 * @param shell Shell instance.
 * @param argc  Arguments count.
 * @param argv  Arguments.
 *
 * @retval 0 Successful command execution.
 * @retval 1 Help printed and command not executed.
 * @retval -EINVAL Argument validation failed.
 * @retval -ENOEXEC Command not executed.
 */
typedef int (*shell_cmd_handler)(const struct shell *shell,
				 size_t argc, char **argv);
```

and only values, 0, 1, -EINVAL and -ENOEXEC should be expected as return values?

nordic-krch reviewed (dismissed):
Looks ok for me. Don't have opinion on retcode vs retval.

jakub-uC commented:
@carlescufi I am aware about this but shell was introduced before these rules. Every time I implement something new or fix a shell I try to improve the old implementation. I can go through the whole thing and fix everything in one PR to not cause further confusion.

jfischer-no commented on subsys/shell/shell.c:
```diff
@@ -1280,22 +1280,14 @@ static void shell_signal_handle(const struct shell *sh,
 		k_poll_signal_reset(sig);
-=======
-		k_poll_signal_reset(signal);
->>>>>>> shell: fix MISRA 5.7 violations on `struct shell`
```
Please drop last two commits.

hsearle replied:
I apologize for my lack of understanding about git. I think my last two commits were in response to the need to rebase my branch onto the main branch. What would be the best way to drop them, clean up the git history, and still have it able to merge into main?

jfischer-no replied:
IMO the best way for you is to squash into one commit, e.g. use `git rebase -i HEAD~4` and edit the list,

```
pick b2c1dfd9f39 Shell: Add get return code command.
pick 75a7a1d6e4c Shell: add retrieve return value via shell API
pick d5a51e0b3b2 shell: fix MISRA 5.7 violations on `struct shell`
pick 6e710d90027 Shell: add retrieve return value via shell API

# Rebase von 546ea7f977f..6e710d90027 auf 546ea7f977f (4 Kommandos)
```

to be

```
pick b2c1dfd9f39 Shell: Add get return code command.
s 75a7a1d6e4c Shell: add retrieve return value via shell API
s d5a51e0b3b2 shell: fix MISRA 5.7 violations on `struct shell`
s 6e710d90027 Shell: add retrieve return value via shell API

# Rebase von 546ea7f977f..6e710d90027 auf 546ea7f977f (4 Kommandos)
```

fix the commit message, git log:
```

commit 8f85799bc8e74bf16b4dd986707b6edf3598774c (HEAD -> 56130)
Author: Hunter Searle <hsearle@xes-inc.com>
Date:   Wed Mar 22 16:20:46 2023 -0500

    Shell: Add get return code command.
    
    This new commmand, and corresponding update to command
    execution, enables the user to retrieve the return code
    of the most recently run shell command. This is useful both
    for users wanting specific information beyond the print
    statements in a command, but especially for automated users
    that prefer a simple return code to parsing human readable text.
    This was tested using all default shell commands, as well as
    eeprom, flash, and i2c, where specific errors could be forced.
    In all cases, the correct return value was returned by the new
    retval command.
    
    Signed-off-by: Hunter Searle <hsearle@xes-inc.com>

commit 546ea7f977f278b2eb212013062bc040c6dc0faa
....

```

optional check the diff to you remove branch (git diff refs/pull/upstream/56130), and force push.

https://git-scm.com/docs/git-rebase#Documentation/git-rebase.txt---interactive

hsearle replied:
Thanks for the guidance.

jfischer-no replied:
Okay, in any case, there is a backup for you https://github.com/jfischer-no/zephyr/commit/8f85799bc8e74bf16b4dd986707b6edf3598774c.patch

carlescufi commented:
@hsearle please use `git rebase -i` or `git cherry-pick` to remove the merge commit. If you are in need of assistance please join Discord and ping us in the #pr-help channel.

hsearle committed:
Shell: Edit shell to store/retrieve return values

This change to the shell API enables the user to retrieve the
return code of the most recently run shell command. This is
useful both for users wanting specific information beyond the print
statements in a command, but especially for automated users
that prefer a simple return code to parsing human readable text.
This was tested using all default shell commands, as well as
eeprom, flash, and i2c, where specific errors could be forced.
In all cases, the correct return value was returned by the new
retval command.

Signed-off-by: Hunter Searle <hsearle@xes-inc.com>

cfriedt commented on include/zephyr/shell/shell.h:
```diff
@@ -1239,6 +1240,15 @@ int shell_obscure_set(const struct shell *sh, bool obscure);
  */
 int shell_mode_delete_set(const struct shell *sh, bool val);
 
+/**
```
Please surround this with
```cpp
#ifdef CONFIG_SHELL_CMDS_RETURN_VALUE
// ...
#endif
```

cfriedt commented on include/zephyr/shell/shell.h:
```diff
@@ -1239,6 +1240,15 @@ int shell_obscure_set(const struct shell *sh, bool obscure);
  */
 int shell_mode_delete_set(const struct shell *sh, bool val);
 
+/**
+ * @brief Retrieve return value of most recently executed shell command.
+ *
+ * @param[in] shell Pointer to the shell instance
+ *
+ * @retval return value of previous command
+ */
+int shell_get_return_value(const struct shell *sh);
```
Please surround this with
```cpp
#ifdef CONFIG_SHELL_CMDS_RETURN_VALUE
// ...
#endif
```

cfriedt commented on subsys/shell/shell.c:
```diff
@@ -1001,7 +1001,7 @@ static void state_collect(const struct shell *sh)
 				} else {
 					/* Command execution */
-					(void)execute(sh);
+					sh->ctx->ret_val = execute(sh);
```
Please surround this with
```cpp
#ifdef CONFIG_SHELL_CMDS_RETURN_VALUE
// ...
#endif
```

cfriedt commented on subsys/shell/shell.c:
```diff
@@ -1670,6 +1670,15 @@ int shell_use_vt100_set(const struct shell *sh, bool val)
 	return (int)z_flag_use_vt100_set(sh, val);
 }
 
+int shell_get_return_value(const struct shell *sh)
+{
+	if (sh == NULL) {
+		return -EINVAL;
+	}
+
+	return z_shell_get_return_value(sh);
+}
```
Please surround this with
```cpp
#ifdef CONFIG_SHELL_CMDS_RETURN_VALUE
// ...
#endif
```

cfriedt commented on subsys/shell/shell_cmds.c:
```diff
@@ -402,6 +403,15 @@ static int cmd_resize(const struct shell *sh, size_t argc, char **argv)
 	return 0;
 }
 
+static int cmd_get_retval(const struct shell *sh, size_t argc, char **argv)
+{
+	ARG_UNUSED(argc);
+	ARG_UNUSED(argv);
+
+	shell_print(sh, "%d", shell_get_return_value(sh));
+	return 0;
+}
+
```
Please surround this with
```cpp
#ifdef CONFIG_SHELL_CMDS_RETURN_VALUE
// ...
#endif
```

cfriedt commented on subsys/shell/shell_cmds.c:
```diff
@@ -498,3 +508,5 @@ SHELL_COND_CMD_ARG_REGISTER(CONFIG_SHELL_CMDS_RESIZE, resize, &m_sub_resize,
 SHELL_COND_CMD_ARG_REGISTER(CONFIG_SHELL_CMDS_SELECT, select, NULL,
 			    SHELL_HELP_SELECT, cmd_select, 2,
 			    SHELL_OPT_ARG_CHECK_SKIP);
+SHELL_COND_CMD_ARG_REGISTER(SHELL_CMDS_RETURN_VALUE, retval, NULL,
```
Please surround this with
```cpp
#ifdef CONFIG_SHELL_CMDS_RETURN_VALUE
// ...
#endif
```

cfriedt commented on subsys/shell/shell_ops.c:
```diff
@@ -485,6 +485,11 @@ void z_shell_vt100_colors_restore(const struct shell *sh,
 	vt100_bgcolor_set(sh, color->bgcol);
 }
 
+static int z_shell_get_return_value(const struct shell *sh)
+{
+	return sh->ctx->ret_val;
+}
```
Please surround this with
```cpp
#ifdef CONFIG_SHELL_CMDS_RETURN_VALUE
// ...
#endif
```

cfriedt commented on include/zephyr/shell/shell.h:
```diff
@@ -823,6 +823,7 @@ struct shell_ctx {
 
 	struct k_mutex wr_mtx;
 	k_tid_t tid;
+	int ret_val;
```
Please surround this with
```cpp
#ifdef CONFIG_SHELL_CMDS_RETURN_VALUE
// ...
#endif
```

cfriedt reviewed (dismissed):
`shell_get_return_value()` should depend on `CONFIG_SHELL_CMDS_RETURN_VALUE`

hsearle commented:
> `shell_get_return_value()` should depend on `CONFIG_SHELL_CMDS_RETURN_VALUE`

In an early discussion, we decided that this functionality could be useful outside of the shell command context. That is why I created the functions in shell.c/shell_ops.c instead of implementing it directly in the command. Given this, I don't think we should make it depend on `CONFIG_SHELL_CMDS_RETURN_VALUE`

cfriedt commented:
> > `shell_get_return_value()` should depend on `CONFIG_SHELL_CMDS_RETURN_VALUE`
> 
> In an early discussion, we decided that this functionality could be useful outside of the shell command context. That is why I created the functions in shell.c/shell_ops.c instead of implementing it directly in the command. Given this, I don't think we should make it depend on `CONFIG_SHELL_CMDS_RETURN_VALUE`

My mistake - I think I misunderstood the purpose of `CONFIG_SHELL_CMDS_RETURN_VALUE`

cfriedt commented:
@hsearle - you might need to rebase and fix failing checks.

cfriedt commented:
There may be some CI issues right now on GitHub's side. Need to wait until those are resolved. When CI passes, I'll approve.

carlescufi commented:
@hsearle there is a documentation build failure now, you need to update the API doc after having changed the param name.

```
/home/runner/work/zephyr/zephyr/include/zephyr/shell/shell.h:1244: argument 'shell' of command @param is not found in the argument list of shell_get_return_value(const struct shell *sh)
```

jfischer-no commented on subsys/shell/shell_ops.c:
```diff
@@ -485,6 +485,11 @@ void z_shell_vt100_colors_restore(const struct shell *sh,
 	vt100_bgcolor_set(sh, color->bgcol);
 }
 
+static int z_shell_get_return_value(const struct shell *sh)
```
This is not right and has not even been tested, 
"warning: implicit declaration of function 'z_shell_get_return_value';"
"warning: 'z_shell_get_return_value' defined but not used",
it builds because of "SHELL_COND_CMD_ARG_REGISTER(SHELL_CMDS_RETURN_VALUE,", where `SHELL_CMDS_RETURN_VALUE` should be `CONFIG_SHELL_CMDS_RETURN_VALUE`.

Necessary changes:
```diff
diff --git a/subsys/shell/shell_ops.c b/subsys/shell/shell_ops.c
index 4ae957eed3c..a60f9de7acd 100644
--- a/subsys/shell/shell_ops.c
+++ b/subsys/shell/shell_ops.c
@@ -485,7 +485,7 @@ void z_shell_vt100_colors_restore(const struct shell *sh,
        vt100_bgcolor_set(sh, color->bgcol);
 }
 
-static int z_shell_get_return_value(const struct shell *sh)
+int z_shell_get_return_value(const struct shell *sh)
 {
        return sh->ctx->ret_val;
 }
diff --git a/subsys/shell/shell_ops.h b/subsys/shell/shell_ops.h
index 39b66014360..d0e7fef6929 100644
--- a/subsys/shell/shell_ops.h
+++ b/subsys/shell/shell_ops.h
@@ -345,6 +345,8 @@ static inline void z_shell_vt100_colors_store(const struct shell *sh,
 void z_shell_vt100_colors_restore(const struct shell *sh,
                                  const struct shell_vt100_colors *color);
 
+int z_shell_get_return_value(const struct shell *sh);
+
 /* This function can be called only within shell thread but not from command
  * handlers.
  */
```

hsearle replied:
You are correct, this was not tested after my last git rebase/history rewrite. I moved the function to be a static inline function in the header. It now works correctly. Build tested for mpfs_icicle.

jfischer-no commented on subsys/shell/shell_cmds.c:
```diff
@@ -498,3 +508,5 @@ SHELL_COND_CMD_ARG_REGISTER(CONFIG_SHELL_CMDS_RESIZE, resize, &m_sub_resize,
 SHELL_COND_CMD_ARG_REGISTER(CONFIG_SHELL_CMDS_SELECT, select, NULL,
 			    SHELL_HELP_SELECT, cmd_select, 2,
 			    SHELL_OPT_ARG_CHECK_SKIP);
+SHELL_COND_CMD_ARG_REGISTER(SHELL_CMDS_RETURN_VALUE, retval, NULL,
```
Must be `SHELL_COND_CMD_ARG_REGISTER(CONFIG_SHELL_CMDS_RETURN_VALUE, retval, NULL,`

hsearle replied:
Edit made and retested.

jfischer-no commented on subsys/shell/Kconfig:
```diff
@@ -251,6 +251,14 @@ config SHELL_AUTOSTART
 	help
 	  If enabled, shell will be automatically started.
 
+config SHELL_CMDS_RETURN_VALUE
```
This option is not used anywhere, see comment below.

hsearle replied:
Edit made and retested.

jfischer-no reviewed (changes requested):
Changes not tested, all those involved in review have failed.

hsearle commented on subsys/shell/Kconfig:
```diff
@@ -251,6 +251,14 @@ config SHELL_AUTOSTART
+	default y
+	help
+	  Enables the retval to retrieve the return value from the most recently executed command.
+	  Enabled by default.
```
Note removed.

hsearle commented on subsys/shell/Kconfig:
```diff
@@ -251,6 +251,14 @@ config SHELL_AUTOSTART
+	depends on SHELL_CMDS
+	default y
+	help
+	  Enables the retval to retrieve the return value from the most recently executed command.
```
This comment is unhelpful. I don't know what issue you are referring to or how to fix it. I shortened the bool value and changed the help text to keep the width of the longest line under 70 chars. Hopefully those are sufficient to address your concerns.

hsearle commented:
> Changes not tested, all those involved in review have failed.

Thank you for finding these issues. It was certainly an oversight on my part, and I'm glad that you caught it before the changes were integrated.

carlescufi commented:
Thanks for your patience and perseverance @hsearle

hsearle commented:
> Thanks for your patience and perseverance @hsearle

This was my first MR to a public repo. Thanks to everyone for helping me through the process.
